code/game/mecha/mech_fabricator.dm

```diff
@@ -281,151 +320,103 @@
-				</script>
-
-				<table style='width: 100%;'>
-				<tr>
-				<td style='width: 65%; padding-right: 10px;'>
-				[left_part]
-				</td>
-				<td style='width: 35%; background: #000;' id='queue'>
-				[list_queue()]
-				</td>
-				<tr>
-				</table>"}
-	var/datum/browser/popup = new(user, "mecha_fabricator", name, 1000, 600)
-	popup.set_content(dat)
-	popup.open(0)
-	onclose(user, "mecha_fabricator")
-	return
-
 /obj/machinery/mecha_part_fabricator/Topic(href, href_list)
 	if(..())
 		return 1
-	var/datum/topic_input/afilter = new /datum/topic_input(href,href_list)
-	if(href_list["part_set"])
-		var/tpart_set = afilter.getStr("part_set")
-		if(tpart_set)
-			if(tpart_set=="clear")
-				part_set = null
-			else
-				part_set = tpart_set
-				screen = "parts"
-	if(href_list["part"])
-		var/T = afilter.getStr("part")
-		for(var/v in files.known_designs)
-			var/datum/design/D = files.known_designs[v]
+
+	if(href_list["menu"]) // Main menu
+		screen = MECHFAB_MAIN_MENU
+
+	if(href_list["category"]) // Inside a "category": cyborg parts, medical etc.
+		selected_category = href_list["category"]
+		screen = MECHFAB_CATEGORY_MENU
+
+	if(href_list["search"])
+		if(href_list["to_search"])
+			temp_search = href_list["to_search"]
+		if(!temp_search)
+			return
+		matching_designs.Cut()
+
+		for(var/I in files.known_designs)
+			var/datum/design/D = files.known_designs[I]
 			if(D.build_type & fabricator_type)
-				if(D.id == T)
-					if(!processing_queue)
-						build_part(D)
-					else
-						add_to_queue(D)
-					break
+				if(findtext(D.name, temp_search))
+					matching_designs.Add(D)
+		screen = MECHFAB_SEARCH_MENU
+
 	if(href_list["add_to_queue"])
-		var/T = afilter.getStr("add_to_queue")
-		for(var/v in files.known_designs)
-			var/datum/design/D = files.known_designs[v]
+		var/design = href_list["add_to_queue"]
+		for(var/I in files.known_designs)
+			var/datum/design/D = files.known_designs[I]
 			if(D.build_type & fabricator_type)
-				if(D.id == T)
+				if(D.id == design)
 					add_to_queue(D)
 					break
-		return update_queue_on_page()
+
+	if(href_list["category_add_all"])
+		for(var/I in files.known_designs)
+			var/datum/design/D = files.known_designs[I]
+			if(D.build_type & fabricator_type && selected_category in D.category)
+				add_to_queue(D)
+
+	if(href_list["main_menu_add_all"])
+		selected_category = href_list["main_menu_add_all"]
+		for(var/I in files.known_designs)
+			var/datum/design/D = files.known_designs[I]
+			if(D.build_type & fabricator_type && selected_category in D.category)
+				add_to_queue(D)
+		selected_category = null
+
 	if(href_list["remove_from_queue"])
-		remove_from_queue(afilter.getNum("remove_from_queue"))
-		return update_queue_on_page()
-	if(href_list["partset_to_queue"])
-		add_part_set_to_queue(afilter.get("partset_to_queue"))
-		return update_queue_on_page()
+		var/index = text2num(href_list["remove_from_queue"])
+		remove_from_queue(index)
+
 	if(href_list["process_queue"])
-		spawn(0)

```

Comment:
Removing `spawn(0)` does not look right...
Please revert, or explain why this is safe.

Reply:
If I understand it right, spawn spawns a separate process at that point after the wait time. Spawn(0) makes it so the rest doesn't wait for this stuff to finish I think.

Reply:
Hm, I don't quite understand this yet so I'm going to add it back in.